examples/permissioned-token/assets/permissions.py

```diff
@@ -0,0 +1,129 @@
+        )),
+
+        # Save max_tokens count in global state (default = 100, during ssc deploy)
+        App.globalPut(max_tokens, Int(100)),
+
+        # Initialize whitelisted accounts counter to 0
+        App.globalPut(whitelist_count, Int(0)),
+
+        # Save controller application index in global state
+        App.globalPut(controller_app_id, Btoi(Txn.application_args[0])),
+        Return(Int(1))
+    ])
+
+    # fetch permissions manager, token_id from controller's global state (using foreignApps)
+    permission_manager = App.globalGetEx(Int(1), Bytes("permissions_manager"))
+    token_id = App.globalGetEx(Int(1), Bytes("token_id"))
+
+    # add an account to be whitelisted (token transfer is rejected if account is not whitelisted)
+    add_whitelist = Seq([
+        permission_manager,
+        token_id,
+        Assert(And(
+		Txn.application_args.length() == Int(2),
+		Txn.accounts.length() == Int(1),
+		Txn.rekey_to() == Global.zero_address(),
+		#Txn.applications.length() ==  Int(1), [TEALv3]
+
+		# verify from global state whether controller application passed is the valid one
+		# note: with tealv3 we can just use txn.applications[0] instead of passing an an appArg
+		Btoi(Txn.application_args[1]) == App.globalGet(controller_app_id),
+
+		# first verify correct token value is passed
+		# token_id.hasValue(),
+			# Txn.assets[0] == token_id.value(), [TEALv3]
+
+		# then verify txn sender is the token manager
+		permission_manager.hasValue(),
+		Txn.sender() == permission_manager.value(),
+        )),
+
+        If(
+            # If account is not already whitelisted, update the counter and continue, return otherwise
+            App.localGet(Int(1), Bytes("whitelisted")) == Int(0),
+            App.globalPut(whitelist_count, App.globalGet(whitelist_count) + Int(1)), # only evaluated if above is true
+            Return(Int(1)) # else return
+        ),
+
+        # finally update txn.accounts[1] local state to set whitelisted status as true(1)
+        App.localPut(Int(1), Bytes("whitelisted"), Int(1)),
+        Return(Int(1))
+    ])
+
+    asset_balance = AssetHolding.balance(Int(1), Gtxn[1].xfer_asset())
+    transfer_token = Seq([
+        asset_balance, # load asset_balance of asset_receiver from store
+        Assert(And(
+		Txn.application_args.length() == Int(1),
+		Global.group_size() >= Int(3),
+		Txn.sender() == Gtxn[1].asset_sender(),
+
+		# verify tx.accounts[1] of current_tx should be same as asset receiver
+		Txn.accounts[1] == Gtxn[1].asset_receiver(),
+
+		# rule 1 - check balance of receiver after receiving token <= 100(max_tokens)
+		asset_balance.hasValue(),
+		asset_balance.value() <= App.globalGet(max_tokens),
+
+		# rule 2 - [from, to] accounts must be whitelisted
+		# NOTE: Int(0) == Txn.Sender(), Int(1) == Txn.accounts[1]
+		App.localGet(Int(0), Bytes("whitelisted")) == true, # from account must be whitelisted
+		App.localGet(Int(1), Bytes("whitelisted")) == true  # to account must be whitelisted
```

Comment:
What if a user will use a different address here (different than a "true" recipient of the transfer)?
can we extract the `to` address from the `Gtxn[1]`?

Reply:
> What if a user will use a different address here (different than a "true" recipient of the transfer)?

We are verifying the `to` address to be asset_receiver before whitelisting: [https://github.com/scale-it/algo-builder/pull/292/files/7bf3c0c34f54f5bdd3a3cd5dc3ff398c26392441#diff-ce9fa1b7e02d1a7a20ef56e646823bb78a55dd88724bad14ecd27e19297b3917R87](https://github.com/scale-it/algo-builder/pull/292/files/7bf3c0c34f54f5bdd3a3cd5dc3ff398c26392441#diff-ce9fa1b7e02d1a7a20ef56e646823bb78a55dd88724bad14ecd27e19297b3917R87) 

> can we extract the to address from the Gtxn[1]?

`App.localGet(..)` expects first arg as index in *AppAccounts* array.